massa-models/src/denunciation.rs

```diff
@@ -0,0 +1,628 @@
+        buffer.extend(value.pub_key.to_bytes());
+        let denunciation_kind = value.is_for_block() as u8;
+        buffer.extend([denunciation_kind]);
+
+        match value.proof.as_ref() {
+            DenunciationProof::Endorsement(ed) => {
+                self.u32_serializer.serialize(&ed.index, buffer)?;
+                buffer.extend(ed.signature_1.to_bytes());
+                buffer.extend(ed.hash_1.to_bytes());
+                buffer.extend(ed.signature_2.to_bytes());
+                buffer.extend(ed.hash_2.to_bytes());
+            }
+            DenunciationProof::Block(bd) => {
+                buffer.extend(bd.signature_1.to_bytes());
+                buffer.extend(bd.hash_1.to_bytes());
+                buffer.extend(bd.signature_2.to_bytes());
+                buffer.extend(bd.hash_2.to_bytes());
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Deserializer for Denunciation
+pub struct DenunciationDeserializer {
+    slot_deserializer: SlotDeserializer,
+    index_deserializer: U32VarIntDeserializer,
+    hash_deserializer: HashDeserializer,
+    sig_deserializer: SignatureDeserializer,
+    pub_key_deserializer: PublicKeyDeserializer,
+}
+
+impl DenunciationDeserializer {
+    /// Creates a new Denunciation deserializer
+    pub fn new(thread_count: u8, endorsement_count: u32) -> Self {
+        Self {
+            slot_deserializer: SlotDeserializer::new(
+                (Included(0), Included(u64::MAX)),
+                (Included(0), Excluded(thread_count)),
+            ),
+            index_deserializer: U32VarIntDeserializer::new(
+                Included(0),
+                Excluded(endorsement_count),
+            ),
+            hash_deserializer: HashDeserializer::new(),
+            sig_deserializer: SignatureDeserializer::new(),
+            pub_key_deserializer: PublicKeyDeserializer::new(),
+        }
+    }
+}
+
+impl Deserializer<Denunciation> for DenunciationDeserializer {
+    fn deserialize<'a, E: ParseError<&'a [u8]> + ContextError<&'a [u8]>>(
+        &self,
+        buffer: &'a [u8],
+    ) -> IResult<&'a [u8], Denunciation, E> {
+        let (rem, (slot, pub_key, is_for_block)) = context(
+            "Failed Denunciation deserialization",
+            tuple((
+                context("Failed slot deserialization", |input| {
+                    self.slot_deserializer.deserialize(input)
+                }),
+                context("Failed pub_key deserialization", |input| {
+                    self.pub_key_deserializer.deserialize(input)
+                }),
+                context("Failed slot deserialization", |input| take(1usize)(input)),
+            )),
+        )
+        .parse(buffer)?;
+
+        // TODO: rework this
+        let is_for_block_ = matches!(is_for_block, [1]);
+
+        let (rem2, proof): (_, DenunciationProof) = match is_for_block_ {
+            true => context(
+                "Failed Block Denunciation deser",
+                tuple((
```

Comment:
You could create a parser for this 4 fields that are commun with endorsements and use it to avoid code duplication

Reply:
Done